module/VuFindSearch/tests/unit-tests/src/VuFindTest/Backend/Solr/ConnectorTest.php

```diff
@@ -140,6 +140,41 @@ public function testSaveThrowsUnknownFormat()
         $conn->write($document, 'unknown', 'update');
     }
 
+    /**
+     * Test caching.
+     *
+     * @return void
+     */
+    public function testCaching()
+    {
+        $conn = $this->createConnector('single-record');
+
+        list(, $expectedBody) = explode("\n\n", $this->response);
+
+        $cache = $this->createMock(\Laminas\Cache\Storage\StorageInterface::class);
+        $cache->expects($this->exactly(3))
+            ->method('getItem')
+            ->willReturnOnConsecutiveCalls(null, $expectedBody, 'foo');
+        $keyConstraint = new \PHPUnit\Framework\Constraint\IsType('string');
+        $cache->expects($this->exactly(1))
+            ->method('setItem')
+            ->with($keyConstraint, $expectedBody)
+            ->will($this->returnValue(true));
+
+        $conn->setCache($cache);
+
+        $resp = $conn->retrieve('id');
+        $this->assertEquals($expectedBody, $resp);
+        $resp = $conn->retrieve('id');
+        $this->assertEquals($expectedBody, $resp);
+        $resp = $conn->retrieve('id');
+        $this->assertEquals('foo', $resp);
+
+        // Make sure that write() doesn't access the cache
```

Comment:
Do we need any more assertions in support of this? Sorry if I'm missing something!

Reply:
The expectations for number of calls to getItem/setItem will fail if write() calls them. I added a comment about that. If you think it's better to make it more explicit, I can still revise.

Reply:
I decided to revise the test to be more explicit. Now it should be easier to read.